File: leetcode/longest-substring-without-repeating-characters.c
/*
topics:
  - string
  - hashmap
  - sliding window
difficulty: medium
*/

int lengthOfLongestSubstring(char* s) {
  int SEEN_LAST_AT[256];
  int ans = 0, left = 0;
  // ugly but cute memset
  memset(SEEN_LAST_AT, -1, 256 * sizeof(int));
  for (int right = 0; s[right]; right++) {
    // set the left to one after the index of this character
    left = fmax(left, SEEN_LAST_AT[s[right]] + 1);
    // update the current subsequence length if greater
    ans = fmax(ans, right - left + 1);
    // update the last index of this character
    SEEN_LAST_AT[s[right]] = right;
  }
  return ans;
}

File: leetcode/longest-substring-without-repeating-characters.ts
/*
topics:
  - string
  - hash table
  - sliding window
difficulty: medium
*/

function lengthOfLongestSubstring(s: string): number {
  let seen: { [id: string]: number; } = {};
  let ans: number = 0, left: number = 0, right: number = 0;
  for (; right < s.length; right++) {
    if (s[right] in seen)
      left = Math.max(left, seen[s[right]]! + 1);
    seen[s[right]] = right;
    ans = Math.max(ans, right - left + 1);
  }
  return ans;
};

// uses linear time